libraries/wallet/wallet.cpp

```diff
@@ -2475,5 +2617,194 @@ std::vector<proposal_api_obj> wallet_api::list_proposals()
+    }
+
+    return atomicswap_contract_result_api_obj(ret);
+}
+
+annotated_signed_transaction wallet_api::atomicswap_redeem(const std::string& from,
+                                                           const std::string& to,
+                                                           const std::string& secret,
+                                                           const bool broadcast)
+{
+    FC_ASSERT(!is_locked());
+
+    atomicswap_redeem_operation op;
+
+    op.from = from;
+    op.to = to;
+    op.secret = secret;
+
+    signed_transaction tx;
+    tx.operations.push_back(op);
+    tx.validate();
+
+    annotated_signed_transaction ret;
+    try
+    {
+        ret = my->sign_transaction(tx, broadcast);
+    }
+    catch (fc::exception& e)
+    {
+        elog("Can't redeem Atomic Swap contract.");
+    }
+
+    return ret;
+}
+
+atomicswap_contract_info_api_obj
+wallet_api::atomicswap_auditcontract(const std::string& from, const std::string& to, const std::string& secret_hash)
+{
+    atomicswap_contract_info_api_obj ret;
+    try
+    {
+        ret = my->_remote_db->get_atomicswap_contract(from, to, secret_hash);
+    }
+    catch (fc::exception& e)
+    {
+        elog("Can't access to Atomic Swap contract.");
+    }
+    return ret;
+}
+
+std::string
+wallet_api::atomicswap_extractsecret(const std::string& from, const std::string& to, const std::string& secret_hash)
+{
+    try
+    {
+        atomicswap_contract_info_api_obj contract_info = atomicswap_auditcontract(from, to, secret_hash);
+
+        FC_ASSERT(!contract_info.secret.empty(), "Contract is not redeemed.");
+
+        return contract_info.secret;
+    }
+    catch (fc::exception& e)
+    {
+        elog("Can't access to Atomic Swap contract secret.");
+    }
+
+    return "";
+}
+
+annotated_signed_transaction wallet_api::atomicswap_refund(const std::string& participant,
+                                                           const std::string& initiator,
+                                                           const std::string& secret_hash,
+                                                           const bool broadcast)
+{
+    FC_ASSERT(!is_locked());
+
+    atomicswap_refund_operation op;
+
+    op.participant = participant;
+    op.initiator = initiator;
+    op.secret_hash = secret_hash;
+
+    signed_transaction tx;
+    tx.operations.push_back(op);
+    tx.validate();
+
+    annotated_signed_transaction ret;
+    try
+    {
+        ret = my->sign_transaction(tx, broadcast);
+    }
+    catch (fc::exception& e)
+    {
+        elog("Can't refund Atomic Swap contract.");
+    }
+
+    return ret;
+}
+
+std::vector<atomicswap_contract_api_obj> wallet_api::get_atomicswap_contracts(const std::string& owner)
+{
+    std::vector<atomicswap_contract_api_obj> result;
+
+    result = my->_remote_db->get_atomicswap_contracts(owner);
+
+    return result;
+}
+
+void wallet_api::exit()
+{
+    std::exit(0);
```

Comment:
this is incorrect call, not all objects will be destroyed.

Reply:
call stop()

Reply:
OS will destroy all of these. It is wallet.

Reply:
OS will not dtor for objects in the stack.